test/extensions_test.dart

```diff
@@ -989,6 +989,52 @@ void main() {
         expect(iterable(['4', '3', '2', '1']).isSorted(), false);
       });
     });
+    group('.choose', () {
+      test('errors', () {
+        expect(() => iterable([]).choose(1), throwsStateError);
+        expect(() => iterable([1]).choose(-1), throwsRangeError);
+        expect(() => iterable([1]).choose(2), throwsStateError);
+      });
+      test('empty', () {
+        var empty = iterable(<int>[]);
+        expect(empty.choose(0), []);
+      });
+      test('single', () {
+        var single = iterable([1]);
+        expect(single.choose(0), []);
+        expect(single.choose(1), [1]);
+      });
+      test('multiple', () {
+        var multiple = iterable([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
+        expect(multiple.choose(0), []);
+        var one = multiple.choose(1);
+        expect(one.length, 1);
+        expect(one.first, inInclusiveRange(1, 10));
+        var seen = <int>{};
+        do {
+          var some = multiple.choose(3);
+          expect(some.length, 3);
```

Comment:
```suggestion
          expect(some, hasLength(3));
```

Reply:
I *really* dislike the `test` package for needing to know a large number of idiosyncratic matchers by name.

Reply:
Yeah, we've had some discussions in the past about writing a nicer API using extension methods where potentially `hasLength` could be surfaced as a suggestion by the analyzer if it was an extension on `Expect<List>` (and we'd need to do the same for other common types with lengths). Then we could do things like `expect(some).hashLength(3)`. It wouldn't help if the first instinct is to write `expect(some.length).equals(3)`, but it might still help some with discoverability.

It's really unfortunate that the existing undiscoverable API is so baked in to `package:test` because it makes it hard to ship something better. I do have some ideas around loosening the coupling to make it feasible to ship alternative APIs, but I don't know if or when we'll be able to get them on OKRs.